sql/atxs/atxs.go

```diff
@@ -440,27 +441,31 @@ func Add(db sql.Executor, atx *types.ActivationTx) error {
 		} else {
 			stmt.BindNull(13)
 		}
+		stmt.BindInt64(14, int64(atx.Weight))
 	}
 
 	_, err := db.Exec(`
 		insert into atxs (id, epoch, effective_num_units, commitment_atx, nonce,
 			 pubkey, received, base_tick_height, tick_count, sequence, coinbase,
-			 validity, prev_id)
-		values (?1, ?2, ?3, ?4, ?5, ?6, ?7, ?8, ?9, ?10, ?11, ?12, ?13)`, enc, nil)
+			 validity, prev_id, weight)
+		values (?1, ?2, ?3, ?4, ?5, ?6, ?7, ?8, ?9, ?10, ?11, ?12, ?13, ?14)`, enc, nil)
 	if err != nil {
 		return fmt.Errorf("insert ATX ID %v: %w", atx.ID(), err)
 	}
 
-	enc = func(stmt *sql.Statement) {
-		stmt.BindBytes(1, atx.ID().Bytes())
-		stmt.BindBytes(2, atx.Blob)
-		stmt.BindInt64(3, int64(atx.Version))
+	return AddBlob(db, atx.ID(), atx.Blob, atx.Version)
+}
+
+func AddBlob(db sql.Executor, id types.ATXID, blob []byte, version types.AtxVersion) error {
```

Comment:
It is still called in `Add()`. I factored it out to avoid creating `types.ActivationTx` in the 21st migration tests to avoid complicating the test (it only needs the blob).

Btw, does the same logic applies to not having `SetUnits()` as a separate function?

Reply:
I think we should move the `AddBlob` function to the package with the state migration and leave the `Add` function unchanged here.

The migration only lives temporarily (we delete the code in the next version anyway) and if `Add` changes (because of a schema change) `AddBlob` used by the migration ***must not*** change (since it still uses the previous schema). Wdyt?

Reply:
Regarding `SetUnits`: yes, if possible we should have that as part of `Add` to avoid having inconsistent DB states, what if `Add` succeeds and `SetUnits` fails?

Reply:
>what if Add succeeds and SetUnits fails?

Well, the requirement (could be documented) is that both are called on the same `tx` (that's what happens in tax handlers). In case any of an error the TX is not commited.

Reply:
Modifying `atxs.Add()` is painful because it is used in many places in tests :cry: Probably, we should eventually land with Add function declaration like:

```go
atxs.Add(atx *types.ActivationTx, blob types.AtxBlob, posts map[types.NodeID]uint32) error
```

:bulb: I'd rather not have `blob` and `posts` on the `types.ActivationTx` to not have to query them on `atxs.Get()` and to avoid having semi-initialized ATXs returned. The only reason for `AtxBlob` being a field on `ActivationTx` was painful refactor I think :smiling_face_with_tear:

Reply:
I think we had this discussion before 😅. I would prefer passing the blob as parameter instead of as a (possibly nil) field of the ATX.

Reply:
> Well, the requirement (could be documented) is that both are called on the same tx (that's what happens in tax handlers). In case any of an error the TX is not committed.

Sure but I prefer not having to add comments to functions to explain how they are intended to be used, because the compiler doesn't tell me if they become incorrect 😉

Reply:
I extracted the blob of out `types.ActivationTx` in a separate PR: https://github.com/spacemeshos/go-spacemesh/pull/6093